third_party/move/move-vm/runtime/src/storage/loader.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    config::VMConfig,
+    loader::{Function, LoadedFunctionOwner, Module},
+    logging::expect_no_verification_errors,
+    module_traversal::TraversalContext,
+    storage::module_storage::ModuleStorage,
+    CodeStorage, LoadedFunction,
+};
+use move_binary_format::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::{Location, PartialVMError, PartialVMResult, VMResult},
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    gas_algebra::NumBytes,
+    identifier::IdentStr,
+    language_storage::{ModuleId, TypeTag},
+    vm_status::StatusCode,
+};
+use move_vm_types::{
+    gas::GasMeter,
+    loaded_data::runtime_types::{StructType, Type, TypeBuilder},
+    module_linker_error,
+};
+use std::{collections::BTreeMap, sync::Arc};
+use typed_arena::Arena;
+
+/// V2 implementation of loader, which is stateless - i.e., it does not contain module or script
+/// cache. Instead, module and script storages are passed to all APIs by reference.
+pub(crate) struct LoaderV2 {
+    vm_config: VMConfig,
```

Comment:
Is this info duplicated with the config in the Environment?

Reply:
Yes, kept it for now cause it's a relatively small struct. Otherwise ideally VM should just take a `&'a VMConfig` but I didn't want to fight lifetimes for both V1 and V2 logics. Once we have V2 only, we can do that, I think.